File: backend/models/deal.py
# =============================================================================
# File: deal.py
# Project: Grocery Intelligence Platform
# Author: Anita Woodford
# Description: Defines deal match models for generated weekly grocery deal reports.
# Security Note: This model stores deal calculation references and financial fields only.
# SRS Traceability: Supports SRS v5.0 deal matching, coupon matching, rebate matching, and report generation requirements.
# SDD Traceability: Supports SDD v5.0 backend model validation and generated deal report design.
# =============================================================================

from __future__ import annotations  # Enables forward-compatible type annotations.

from datetime import UTC, date, datetime  # Imports date and timezone-aware datetime support.
from decimal import Decimal  # Imports Decimal for exact currency values.
from typing import Optional  # Imports Optional for nullable reference fields.

from pydantic import ConfigDict, Field  # Imports Pydantic configuration and field helpers.

from backend.models.ad import DealType  # Imports the shared DealType enum instead of redefining it.
from backend.models.base import BaseDocument, PyObjectId  # Imports the shared base document and MongoDB object id type.


class DealMatchItem(BaseDocument):  # Defines one matched deal result inside a weekly deal report.
    """Represents one product-store deal match with coupon, rebate, and reward results."""  # Documents the purpose of the nested model.

    model_config = ConfigDict(extra="forbid", arbitrary_types_allowed=True)  # Rejects unknown fields and allows ObjectId-compatible values.

    product_ref: PyObjectId  # Stores the matched product reference id.
    store_ref: PyObjectId  # Stores the store reference id for the matched deal.
    deal_ref: Optional[PyObjectId] = None  # Stores the optional ad or deal reference id.
    coupon_refs: list[PyObjectId] = Field(default_factory=list)  # Stores coupon reference ids used in the deal.
    rebate_refs: list[PyObjectId] = Field(default_factory=list)  # Stores rebate reference ids used in the deal.
    shelf_price: Decimal  # Stores the regular shelf price before sale pricing.
    register_price: Decimal  # Stores the register price after store sale pricing and coupons.
    oop: Decimal  # Stores the out-of-pocket amount paid at checkout.
    rr_earned: Decimal = Decimal("0.00")  # Stores Walgreens Register Rewards earned.
    ecb_earned: Decimal = Decimal("0.00")  # Stores CVS ExtraBucks earned.
    total_rebates_back: Decimal = Decimal("0.00")  # Stores total rebate value expected back.
    deal_type: DealType  # Stores the shared deal type imported from the ad model.
    is_money_maker: bool = False  # Marks whether rewards and rebates exceed out-of-pocket cost.
    matched_as_substitute: bool = False  # Marks whether the item matched through an acceptable substitute.


class DealModel(BaseDocument):  # Defines the weekly generated deal report model.
    """Represents generated deal matches for one shopping week."""  # Documents the purpose of the top-level model.

    model_config = ConfigDict(extra="forbid", arbitrary_types_allowed=True)  # Rejects unknown fields and allows ObjectId-compatible values.

    week_of: date  # Stores the week date for the generated deal report.
    generated_at: datetime = Field(default_factory=lambda: datetime.now(UTC))  # Stores the UTC generation timestamp.
    matches: list[DealMatchItem] = Field(default_factory=list)  # Stores typed deal match items instead of raw dictionaries.

File: frontend/src/types/deal.ts
/** 
 * File: 
 * Description: T
 * Author: Anita Woodford
 * Created: 2024-06-01
 * License: MIT
 * Copyright (c) 2024 Anita Woodford. All rights reserved.
 * SRS Document: 
 * SDD Document: 
 * Revision History:
 */

import type { BaseDocument, ObjectId, StandardResponse } from './base'; // Imports shared frontend types.

export type DealType = // Defines supported deal classifications.
  | 'STANDARD' // Represents a standard sale deal.
  | 'BOGO' // Represents buy-one-get-one.
  | 'BUY_X_GET_Y' // Represents buy X get Y.
  | 'THRESHOLD' // Represents spend or quantity threshold deal.
  | 'REGISTER_REWARD' // Represents Walgreens Register Reward deal.
  | 'ECB' // Represents CVS ExtraBucks deal.
  | 'SPEND_BOOSTER' // Represents a spend booster deal.
  | 'MANUFACTURER_COUPON' // Represents manufacturer coupon deal.
  | 'COMBO_LOCO' // Represents H-E-B Combo Loco.
  | 'BASKET_COUPON' // Represents basket-level coupon.
  | 'PERCENT_OFF'; // Represents percent-off deal.

export interface DealItem { // Defines one item inside an ad or deal.
  item_name: string; // Stores deal item name.
  brand?: string | null; // Stores brand when available.
  size?: string | null; // Stores size when available.
  sale_price?: string | null; // Stores sale price as Decimal-safe string.
  regular_price?: string | null; // Stores regular price as Decimal-safe string.
  buy_qty?: number | null; // Stores required buy quantity.
  deal_type: DealType; // Stores deal type.
  store_coupon_amount?: string | null; // Stores store coupon amount.
  store_coupon_desc?: string | null; // Stores store coupon description.
  manufacturer_coupon_amount?: string | null; // Stores manufacturer coupon amount.
  manufacturer_coupon_desc?: string | null; // Stores manufacturer coupon description.
  register_reward?: string | null; // Stores Register Reward amount.
  raw_text?: string | null; // Stores original deal text when available.
}

export interface AdModel extends BaseDocument { // Defines one weekly ad record.
  store_ref: ObjectId | string; // Stores linked store id or Phase 1 store key.
  start_date: string; // Stores ad start date as YYYY-MM-DD.
  end_date: string; // Stores ad end date as YYYY-MM-DD.
  source_type?: string | null; // Stores source type when available.
  raw_input_ref?: ObjectId | null; // Stores raw input reference when available.
  items: DealItem[]; // Stores deal items in the ad.
}

export interface DealMatchItem { // Defines one matched deal report item.
  product_ref?: ObjectId | null; // Stores matched product reference when available.
  store_ref: ObjectId | string; // Stores matched store reference.
  deal_ref?: ObjectId | null; // Stores matched deal reference when available.
  coupon_refs?: ObjectId[]; // Stores related coupon references.
  rebate_refs?: ObjectId[]; // Stores related rebate references.
  shelf_price: string; // Stores shelf price as Decimal-safe string.
  register_price: string; // Stores register price as Decimal-safe string.
  oop: string; // Stores out-of-pocket value as Decimal-safe string.
  rr_earned?: string | null; // Stores Register Rewards earned separately.
  ecb_earned?: string | null; // Stores ExtraBucks earned separately.
  reward_value?: string | null; // Stores non-register reward value separately.
  total_rebates_back?: string | null; // Stores total rebate value separately.
  deal_type: DealType; // Stores matched deal type.
  is_money_maker: boolean; // Stores whether deal is a money maker.
  matched_as_substitute: boolean; // Stores whether matched as substitute.
}

export interface DealMatchModel extends BaseDocument { // Defines a generated weekly deal match document.
  week_of: string; // Stores week start date as YYYY-MM-DD.
  generated_at: string; // Stores generation timestamp.
  matches: DealMatchItem[]; // Stores matched deal items.
}

export interface CreateAdRequest { // Defines request body for manual ad entry.
  store_ref: ObjectId | string; // Sends store id or Phase 1 store key.
  start_date: string; // Sends ad start date.
  end_date: string; // Sends ad end date.
  source_type?: string | null; // Sends source type when available.
  items: DealItem[]; // Sends ad deal items.
}

export type AdResponse = StandardResponse<AdModel>; // Defines one-ad API response.
export type AdListResponse = StandardResponse<AdModel[]>; // Defines ad-list API response.
export type DealMatchResponse = StandardResponse<DealMatchModel>; // Defines deal-match API response.